lib/history/sqlite.go

```diff
@@ -0,0 +1,379 @@
+// can hold a specified amount of events and uses a FIFO eviction policy.
+// Timeline events are stored in a local sqlite database.
+//
+// Implements Timeline
+type SQLiteTimeline struct {
+	// clock is used to record event timestamps.
+	clock clockwork.Clock
+	// capacity specifies the max number of events that can be stored in the timeline.
+	capacity int
+	// size specifies the current number of events stored in the timeline.
+	size int
+	// database points to underlying sqlite database.
+	database *sqlx.DB
+	// lastStatus holds the last recorded cluster status.
+	lastStatus *pb.SystemStatus
+	// mu locks timeline access.
+	mu sync.Mutex
+}
+
+// SQLiteTimelineConfig defines SQLiteTimeline configuration.
+type SQLiteTimelineConfig struct {
+	// DBPath specifies the database location.
+	DBPath string
+	// Capacity specifies the max number of events that can be stored in the timeline.
+	Capacity int
+	// Clock will be used to record event timestamps.
+	Clock clockwork.Clock
+}
+
+// NewSQLiteTimeline initializes and returns a new SQLiteTimeline with the
+// specified configuration.
+func NewSQLiteTimeline(config SQLiteTimelineConfig) (*SQLiteTimeline, error) {
+	database, err := initSQLite(config.DBPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &SQLiteTimeline{
+		clock:    config.Clock,
+		capacity: config.Capacity,
+		size:     0,
+		database: database,
+		// TODO: store and recover lastStatus in case satellite agent restarts.
+		lastStatus: nil,
+	}, nil
+}
+
+// initSQLite initializes connection to database provided by dbPath and
+// initializes `events` table.
+func initSQLite(dbPath string) (*sqlx.DB, error) {
+	// TODO: Store sql db connection timeout as const
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second*5)
+	defer cancel()
+
+	database, err := sqlx.ConnectContext(ctx, "sqlite3", dbPath)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if _, err := database.ExecContext(ctx, createTableEvents); err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return database, nil
+}
+
+// RecordStatus records the differences between the previously stored status and
+// the provided status.
+func (t *SQLiteTimeline) RecordStatus(ctx context.Context, status *pb.SystemStatus) (err error) {
+	events := diffCluster(t.clock, t.lastStatus, status)
+	if len(events) == 0 {
+		return nil
+	}
+
+	t.mu.Lock()
+	defer t.mu.Unlock()
+
+	tx, err := t.database.BeginTx(ctx, nil)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	defer func() {
+		// The rollback will be ignored if the tx has already been committed.
+		if err == nil {
+			return
+		}
+		if err := tx.Rollback(); err != nil {
+			log.WithError(err).Error("Failed to rollback sql transaction.")
+		}
+	}()
+
+	if err = t.insertEvents(ctx, tx, events); err != nil {
+		return trace.Wrap(err, "failed to insert events")
+	}
+
+	if t.size+len(events) > t.capacity {
+		if err = t.evictEvents(ctx, tx); err != nil {
```

Comment:
To be frank, I would decouple the process of inserting events from the process of evicting events. Both "structurally", i.e. have separate methods on your timeline object that do that (and as a part of the interface), and "procedurally", i.e. do them in separate transactions. This will give us more flexibility, for example, if we decide to move cleanup to a separate routine, or to treat the eviction as not the critical error. Which btw I don't think it should be a critical error - if we failed to remove extra events - why rollback everything?